rotkehlchen/chain/evm/decoding/compound/v3/balances.py

```diff
@@ -0,0 +1,137 @@
+        underlying_token: dict['ChecksumEvmAddress', 'EvmToken'] = {}
+        for address, events in self.addresses_with_activity(
+            products=None,
+            event_types={(HistoryEventType.RECEIVE, HistoryEventSubType.GENERATE_DEBT)},
+        ).items():
+            for event in events:
+                if event.address is not None:
+                    try:
+                        _compound_token = EvmToken(evm_address_to_identifier(
+                            address=event.address,
+                            chain_id=self.evm_inquirer.chain_id,
+                            token_type=EvmTokenKind.ERC20,
+                        ))
+                        if len(_compound_token.underlying_tokens) == 0:
+                            log.error(f'No underlying token found for {_compound_token!s}.')
+                            continue
+
+                        _underlying_token = EvmToken(evm_address_to_identifier(
+                            address=_compound_token.underlying_tokens[0].address,
+                            chain_id=self.evm_inquirer.chain_id,
+                            token_type=EvmTokenKind.ERC20,
+                        ))
+                    except (UnknownAsset, WrongAssetType) as e:
+                        log.error(
+                            f"Failed to resolve compound v3 borrow event's token and/or its "
+                            f'underlying token in {event.tx_hash.hex()} due to {e!s}',
+                        )
+                        continue
+
+                    unique_borrows[_compound_token].add(address)
+                    underlying_token[event.address] = _underlying_token
+
+        return unique_borrows, underlying_token
+
+    def query_balances(self) -> 'BalancesSheetType':
+        """
+        Query liabilities for Compound v3 open positions and return them. The assets are handled
+        by the wrapped compound v3 tokens, so those are not queried here.
+
+        It calls the Compound token contracts to get the borrow balances of the addresses and
+        tokens whose borrow history events are found in the userDB.
+        """
+        balances: BalancesSheetType = defaultdict(BalanceSheet)
+        unique_borrows, underlying_token = self._extract_unique_borrowed_tokens()
+        if len(unique_borrows) == 0:
+            return balances
+
+        # prepare the arguments for the contract calls and the encoded call
+        calls: list[tuple['ChecksumEvmAddress', str]] = []
+        calls_arguments: list[tuple['ChecksumEvmAddress', 'EvmToken']] = []
+        address_to_contract: dict['ChecksumEvmAddress', 'EvmContract'] = {}
+        for compound_asset, user_addresses in unique_borrows.items():
+            address_to_contract[compound_asset.evm_address] = self.evm_inquirer.contracts.contract(compound_asset.evm_address)  # noqa: E501
+            for user_address in user_addresses:
```

Comment:
the abi is the same for all the tokens right? In other part of the code we use

```python
rate = self.ethereum.call_contract(
          contract_address=address,
          abi=self.ethereum.contracts.abi('CTOKEN'),
          method_name=method_name,
      )
```

since the contract address is not needed when encoding you should be able to initialize it once and then reuse it for the other calls. As you can see you encode borrowBalanceOf and then later you call it for different addresses

Reply:
I was checking the contracts that you added in the sqldiff and is not this. What did you add and where is it needed?

Reply:
I added the token contracts for compound v3, like cUSDCv3. They are used in `self.evm_inquirer.contracts.contract(compound_asset.evm_address)`. But, I see what you mean, I'm removing the token contracts.